README
# JS - Numberals to English

## Documentation
*Changelog in meta/changelog

## Getting started
* Clone, cd code, npm install, npm start

meta/changelog.txt
19/10/2016 - 0.0.1: Generated project, brought in own esLint, changed component structure.
19/10/2016 - 0.0.2: Fixed Sass webpack loader dependency issue, added form component. Set up skeleton for form component with library 'formous'
19/10/2016 - 0.0.3: Fleshed out component structure. Form submittion, clear form, mutating form value.
19/10/2016 - 0.0.4: Component logic cleaned up.
20/10/2016 - 0.0.5: Translation 0-999 OK.
20/10/2016 - 0.0.6: Added sketch for "in between labels" (e.g: "one THOUSAND AND three"). See "addThousands".
20/10/2016 - 0.0.7: Translation OK!
20/10/2016 - 0.0.8: Cleaned up, fixed last minute bugs
20/10/2016 - 0.0.9: Misc. cleanup.
21/10/2016 - 0.1.0: Fixed bug with translation of larger numbers.
21/10/2016 - 0.1.1: More general makeup of the component elements and cleaned up translateInput.
23/10/2016 - 0.1.2: Unit tests and improved structure of helper functions.
23/10/2016 - 0.1.3: Small formatting improvements, added logic for 'zero'.
23/10/2016 - 0.1.4: Fixed console logs for test fail of translation function.
23/10/2016 - 0.1.5: Improved translation comments, cleared code up a little. Fixed edgecases ('00011111' / '000').

code/test/components/FormComponentTest.js
/*eslint-env node, mocha */
/*global expect */
/*eslint no-console: 0*/
'use strict';

import React from 'react';
import { mount } from 'enzyme'
//import sinon from 'sinon';
import Form from 'components/FormComponent/FormComponent.js';
import translationEngine from './../../src/helpers/translateNumber.js'

describe('<FormComponent />', () => {

	it('allows us to set props', () => {
		const wrapper = mount(<Form/>);

		wrapper.setProps({ bar: "foo" });
		expect(wrapper.props().bar).to.equal("foo");
	});

	//This test fails, seems like Formous events are not properly captured during testing. I have written a new test below that tests the translation function itself.

		/*it('simulates click events', () => {
			 const wrapper = mount(<Form />),
				component = wrapper.find('FormComponent'),
				props = component.props();

			//wrapper.find('.FormComponent__inputArabic').simulate('keydown', { which: '1' })
			wrapper.find('.FormComponent__inputArabic').simulate('change', { target: { value: '1' }})
			wrapper.find('.FormComponent__submit').simulate('click');
			expect(wrapper.find('FormComponent').props().fields.arabic.value).to.equal('1');*/


			/**
			 * Keep this as reference
			 */
				/*
					const onButtonClick = sinon.spy();
					const wrapper = shallow(
						<Foo onButtonClick={onButtonClick} />
					);
					expect(onButtonClick).to.have.property('callCount', 1);
					expect(onButtonClick.calledOnce).to.equal(true);
				*/
		//});

	// Do some translation tests and match output with expected value
	it('should translate in an expected manner', () => {
		// Function to test an array of tests
		const callQueries = input => {

			// Prototype query
			let protoQuery = {
				arabic : {
					value : null
				}
			}

			// Results array
			let results = {
				failed : [],
				passed : [],
			};

			// For each item in the array, run a test
			input.forEach(function (item, index) {
				let query = protoQuery;
				query.arabic.value = item.input;

				let translation = translationEngine(query)

				// See if it matches or not...
				if (item.expect === translation.english.value){
					// Catalogue the result
					results.passed.push({
						input: item.input,
						output : translation.english.value,
						expected: item.expect
					})
				}else{
					// Catalogue the result
					results.failed.push({
						input: item.input,
						output : translation.english.value,
						expected: item.expect
					})

					// Log the result
					console.log('');
					console.log('');
					console.log('/**');
					console.log('Sorry, this translation test failed:');
						console.log('	Input: ' + item.input);
						console.log('	Output: ' + translation.english.value);
						console.log('	Expected: ' + item.expect);
				}
			});

			// Log result
			console.log('');
			console.log('');
			console.log('');
			console.log('/***');
			console.log('Translation test results:');
				console.log('	Failed: ' + results.failed.length);
				console.log('	Passed: ' + results.passed.length);
			console.log('');

			// Finish test
			expect(results.failed.length).to.equal(0);
		}

		// Call some queries to test the translation
		callQueries([
			{
				input: '1521100',
				expect: 'one million and five hundred twenty-one thousand and one hundred'
			},
			{
				input: '100',
				expect: 'one hundred'
			},
			{
				input: '5',
				expect: 'five'
			},
			{
				input: '0',
				expect: 'zero'
			},
			{
				input: '2145251255',
				expect: 'two billion and one hundred forty-five million and two hundred fifty-one thousand and two hundred fifty-five'
			},
			{
				input: '23214',
				expect: 'twenty-three thousand and two hundred fourteen'
			},
			{
				input: '98561478',
				expect: 'ninety-eight million and five hundred sixty-one thousand and four hundred seventy-eight'
			}
		]);

	});

});

code/src/helpers/index.js
/*******************************
 * [_index.js]
 * index file for the helpers
 ******************************/

/**
* { Dependencies }
*/
import translateNumberHelper from './translateNumber';
import MutationHelper from './mutation';

/**
* { Function }
*/
const index = (function () {

			/**
			* Support helpers for translation
			*/
			const translation = {
				number : function (input, callback) {
					translateNumberHelper(input, callback);
				}
			};

			/**
			* Support helpers for mutation
			*/
			const mutation = {
				typography : {
					capitaliseFirstLetter : function (input){
						return MutationHelper.typography.capitaliseFirstLetter(input);
					},
					removeWhitespaceDuplicate : function (input){
						return MutationHelper.typography.removeWhitespaceDuplicate(input);
					},
					removeWhitespaceTrailing : function (input){
						return MutationHelper.typography.removeWhitespaceTrailing(input);
					},
					removeWhitespaceTrailingLeading : function (input){
						return MutationHelper.typography.removeWhitespaceTrailingLeading(input);
					}
				}
			};

			return {
				translation : translation,
				mutation : mutation
			};
})();


/**
 * Export
 */
module.exports = index;


code/src/helpers/mutation.js
/*******************************
 * [_mutation.js]
 * Define the helper function for misc. mutatation here
 ******************************/

/**
 * { Dependencies }
 */

 /**
 * { Function }
 */
 const index = (function () {

			/**
			 * { Typography }
			 * Support helpers for typography
			*/
			const typography = {

				// Support to capitalise a result
				capitaliseFirstLetter : function (input) {
					return input.charAt(0).toUpperCase() + input.slice(1);
				},

				removeWhitespaceDuplicate : function (input) {
					return input.replace(/\s+/g, ' ');
				},

				removeWhitespaceTrailing : function (input) {
					return input.replace(/((\s*\S+)*)\s*/, '$1');
				},

				removeWhitespaceTrailingLeading : function (input) {
					console.log(input);
					return input.replace(/\s*((\S+\s*)*)/, '$1');
				}
			};

			return {
				typography : typography
			};
 })();


 /**
  * Export
  */
 module.exports = index;


code/src/helpers/translateInput.js
/*******************************
 * [_translateInput.js]
 * Define the translation code here
 ******************************/

let numberToWords = (input, callback) => {

	const arabic = input.arabic.value;

	/**
	 * Support functions
	 */
		const wordDefinitions = {
			singles : [
				'', 'one', 'two', 'three', 'four',
				'five', 'six', 'seven', 'eight', 'nine',
				'ten', 'eleven', 'twelve', 'thirteen', 'fourteen',
				'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
			],

			tens : [
				'', '', 'twenty', 'thirty', 'forty',
				'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
			],

			thousands(lengthAfter) {
				const x = lengthAfter;

				switch (true) {
					case (x < 3):
						return '';
					case (x >= 3 && x < 5):
						return ' thousand ';
					case (x < 9):
						return ' million ';
					case (x < 12):
						return ' billion ';
					case (x < 15):
						return ' trillion ';
					case (x < 18):
						return ' quadrillion ';
					case (x < 21):
						return ' quintillion ';
					case (x < 24):
						return ' sextillion ';
					case (x < 26):
						return ' octillion ';
					case (x < 29):
						return ' nonillion ';
					case (x > 29):
						return 'Dude, big number. Tone it down.';
					default:
						return '';
				}
			}
		};

		const helpers = {
			reverse(payload) {
				return payload.reverse();
			},

			divideInto(arr, chunkSize) {
				let groups = [], i;

				for (i = 0; i < arr.length; i += chunkSize) {
						groups.push(arr.slice(i, i + chunkSize));
				}
				return groups;
			},

			matchDefinitions([singles, tens, hundreds]) {
				// Array position defines what it is. 0 = singles, 1 = tens, 2 = hundreds

				return [
				// Result for the hundreds
					// If result is 0, return blank
					// Else return the definition for the single digit based on index in array
					hundreds == 0 || hundreds == null ? '' : wordDefinitions.singles[hundreds] + ' hundred ',

				// Result for the singles and tens
					// If single digit is 0, return the definition for the base tens (2 digit) number (e.g: 20, 30)
					// Else return definition of the base tens + '-' + single digit
					// If no arguments, return blank
					singles == 0 ? wordDefinitions.tens[tens] :
					wordDefinitions.tens[tens] && wordDefinitions.tens[tens] + '-' || '',

				// Attach the tens and singles if available, otherwise just take the singles
					wordDefinitions.singles[tens + singles] || wordDefinitions.singles[singles]
				].join(''); //Join them
			},

			addInBetweens(itemCount) {
				/*
				 * Primer: Here we add the "in-between-labels".
				 * To do this, we take the amount of items *after* the current block (passed to us)
				 * This way you input out how big the number is that you're describing
				 */

				// 4. Match the value to a label and return it
				return wordDefinitions.thousands(itemCount);
			}
		};


	/**
	* Process the input
	*/
		/*
		 * Primer
		 * This function takes an input array, reverses it (to free up the first digit) and divides it into blocks of 3.
		 * It then tries to map definitions based on the number that has and adds extra words, see below.
		 * Blocks of three because that makes it easy to translate.
		 * English translations can be thought of as descriptions for blocks of 3 numbers
		 * Aditionally, there are "in-between-labels" such as "thousand and"
		 * E.g.: "400 240" -> "fourhundred *thousand and*  twohundred fourty"
		 * This labels are passed on at the end by calculating how many numbers are after the current block.
		 */

		// 1. Throw the entire input into an array
		let inputArray = [...arabic];

		// 2. Reverse it so that it is spaced into neat blocks and only the first section of the number has gaps in the array
		// (so that we can name it properly when we add thousands)
		inputArray = helpers.reverse(inputArray);

		// 3. Divide it into chunks and fill it to not mess up the mapping later on
		inputArray = helpers.divideInto(inputArray, 3);

		inputArray.map(function (item) {
			while (item.length != 3){
				item.push(null);
			}
		});

		// 4. Map each item to a definition
		let untranslatedArray = inputArray,
			translation = inputArray.map(helpers.matchDefinitions);

		// 5. Reverse it again
		translation = helpers.reverse(translation);

		// 6. Filter empty
		translation = translation.filter(Boolean);

		// 7. Add in-between-labels
		const translationItems = translation.length;

		translation.forEach(function (item, index) {

			let blockIndex = index,
				itemCount = 0;

			// Grab the array as it was before I translated it
			untranslatedArray.forEach(function (item, index) {
				if (index > blockIndex){
					itemCount = itemCount + untranslatedArray[index].length;
				}
			});

			// Plug in the labels
			let label = helpers.addInBetweens(itemCount);

			// If it's the last item in the array, don't add an 'and', otherwise do
			if (index + 1 == translationItems && translationItems != 1){
				//don't do it
			} else {
				label = label + 'and';
			}

			if (label != '') {
				translation[index] = item + label;
			}
		});


		// 8. Clean up
		if (translation.length == 1) {
			// If there is only 1 item in the array, you don't need the "and part".
			translation = translation.join(' ');
			translation = translation.split(' and');
			translation = translation[0];

			// Remove duplicate whitespaces
			translation = translation.replace(/\s+/g, ' ');
		} else {
			translation = translation.join(' ');
			translation = translation.replace(/\s+/g, ' ');
		}

		const result = {
			english : {
				value : translation
			}
		};

		callback(result);
};

export default numberToWords;


code/src/components/FormComponent/FormComponent.js
/*******************************
 * [_FormComponent.js]
 * Define the FormComponent code here
 ******************************/

/* eslint react/prop-types: 0 */

/**
* { Dependencies }
*/
import React from 'react';
import Formous from 'formous';
require('./FormComponent.scss');
import helpers from './../../helpers/index';


/**
 * { Component }
 */

class ErrorText extends React.Component {
	render() {
		return <div style={{color : '#f00' }}>
					{this.props.errorText}
				</div>;
	}
 }

 ErrorText.propTypes = {
	// errorText : React.PropTypes.string
	// Prop validation is already done through Formous
};

class FormComponent extends React.Component {
	constructor(props) {
		super(props);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.handleKeyPress = this.handleKeyPress.bind(this);
		this.mutateComponent = this.mutateComponent.bind(this);
		this.shouldHideTranslationHeader = true;
	}

	componentWillMount() {
	}

	componentDidMount() {
		// console.log(this.props);
	}

	componentWillReceiveProps(nextProps) {
			this.props.setDefaultValues({
				arabic : '',
				english : '',
				shouldHideTranslationHeader : true
			});
	}

	handleSubmit(formStatus, fields) {
		// Initial state
		let fieldState = this.props.fields;

		if (!formStatus.touched) {
			alert('Please fill out the form.');
			return;
		}

		if (!formStatus.valid) {
			alert('Please fill in a number');
			return;
		}

		if (this.props.fields.arabic.value.replace(/ /g, '') == '') {
			alert('Please fill in a number');
			return;
		}

		// No errors found, continue.

		// Translate and mutate result
		helpers.translation.number(fieldState, (result) => this.mutateComponent(result));
	}

	handleKeyPress(e) {
		if (e.key === 'Enter') {
			// Prevent enter key, Formous doesn't seem to like the enter key
			e.preventDefault();
		}
	}

	mutateComponent(payload){
		const typographyHelp = helpers.mutation.typography,
			translation = typographyHelp.capitaliseFirstLetter(payload.english.value);

		// Pass back into the view
		this.setState(this.props.fields.english = {
			value : translation,
			events : this.props.fields.english.events,
			valid : this.props.fields.english.valid
		});

		this.shouldHideTranslationHeader = false;
	}

	render() {

		/*eslint-disable */
		const {
			fields : { arabic, english },
			formSubmit
		} = this.props;
		/*eslint-enable */

		return (
		<div className="FormComponent__outerContainer">
			<div className="FormComponent__title"><h1>Translation, please</h1></div>
			<div className="FormComponent__innerContainer">

				<form onSubmit={formSubmit(this.handleSubmit)}>
					<div className="FormComponent__form">
						<div className="FormComponent__input">
							<div className="FormComponent__arabic" >
								<h4 className="header">Arabic numeral input:</h4>
								<input
									placeholder="Arabic numeral"
									type="text"
									ref="FormComponent__inputArabic"
									className="FormComponent__inputArabic"
									value={arabic.value}
									onKeyPress={this.handleKeyPress}
									{ ...arabic.events }
								/>
								<ErrorText { ...arabic.failProps } />
							</div>

							<hr className={this.shouldHideTranslationHeader ? 'hidden' : ''} />

							<div className="FormComponent__english" >
								<h4 className={this.shouldHideTranslationHeader ? 'hidden header' : 'header'}>
									Your translation:
								</h4>
								<br />

								<p className="FormComponent__outputEnglish">{this.props.fields.english.value}</p>
							</div>
						</div>
					</div>
						<div className="FormComponent__buttons">
							<button className="FormComponent__submit" type="submit">Submit</button>
							<button className="FormComponent__clear" type="button" onClick=
							{(event) => {
								this.shouldHideTranslationHeader = true;
								this.props.clearForm();
							}}>Clear</button>
						</div>
				</form>
			</div>
		</div>
		);
	}
};

FormComponent.propTypes = {
	// Prop validation is already done through Formous
};

const formousOptions = {
	fields : {
		arabic : {
			tests : [
				{
					critical : true,
					failProps : {
						errorText : 'Arabic should be a number'
					},
					test(value) {
						return /^\d*$/.test(value);
					}
				}
			]
		},
		english : {
			tests : [
				{
					critical : true,
					failProps : {
						errorText : 'English should be in letters!'
					},
					test(value) {
						return /^[a-zA-Z-]*$/.test(value);
					}
				},
				{
					critical : true,
					failProps : {
						errorText : 'English should be in letters!'
					},
					test(value) {
						return /^[a-zA-Z-]*$/.test(value);
					}
				}
			]
		}
	}
};

export default Formous(formousOptions)(FormComponent);


code/src/helpers/translateNumber.js
/*******************************
 * [_translateNumber.js]
 * Define the translation code here
 ******************************/
 /**
 * Dependencies
 */

// General helpers
import mutationHelpers from './mutation';


let numberToWords = (input, callback) => {

	/**
	 * Definitions
	 */
	const arabic = input.arabic.value;

	let result = {
		english : {
			value : null
		}
	};

	/**
	 * Support functions for main function down the page
	 */
		const wordDefinitions = {
			singles : [
				'', 'one', 'two', 'three', 'four',
				'five', 'six', 'seven', 'eight', 'nine',
				'ten', 'eleven', 'twelve', 'thirteen', 'fourteen',
				'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
			],

			tens : [
				'', '', 'twenty', 'thirty', 'forty',
				'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
			],

			thousands(lengthAfter) {
				const x = lengthAfter;

				switch (true) {
					case (x < 3):
						return '';
					case (x >= 3 && x < 5):
						return ' thousand ';
					case (x < 9):
						return ' million ';
					case (x < 12):
						return ' billion ';
					case (x < 15):
						return ' trillion ';
					case (x < 18):
						return ' quadrillion ';
					case (x < 21):
						return ' quintillion ';
					case (x < 24):
						return ' sextillion ';
					case (x < 26):
						return ' octillion ';
					case (x < 29):
						return ' nonillion ';
					case (x > 29):
						return 'Dude, big number. Tone it down.';
					default:
						return '';
				}
			}
		};

		// Function specific helpers
		const helpers = {
			reverse(payload) {
				return payload.reverse();
			},

			divideInto(arr, chunkSize) {
				let groups = [], i;

				for (i = 0; i < arr.length; i += chunkSize) {
						groups.push(arr.slice(i, i + chunkSize));
				}
				return groups;
			},

			matchDefinitions([singles, tens, hundreds]) {
				// Array position defines what it is. 0 = singles, 1 = tens, 2 = hundreds

				return [
				// Result for the hundreds
					// If result is 0, return blank
					// Else return the definition for the single digit based on index in array
					hundreds == 0 || hundreds == null ? '' : wordDefinitions.singles[hundreds] + ' hundred ',

				// Result for the singles and tens
					// If single digit is 0, return the definition for the base tens (2 digit) number (e.g: 20, 30)
					// Else return definition of the base tens + '-' + single digit
					// If no arguments, return blank
					singles == 0 ? wordDefinitions.tens[tens] :
					wordDefinitions.tens[tens] && wordDefinitions.tens[tens] + '-' || '',

				// Attach the tens and singles if available, otherwise just take the singles
					wordDefinitions.singles[tens + singles] || wordDefinitions.singles[singles]
				].join(''); //Join them
			},

			addInBetweens(itemCount) {
				/*
				 * Primer: Here we add the "in-between-labels".
				 * To do this, we take the amount of items *after* the current block (passed to us)
				 * This way you input out how big the number is that you're describing
				 */

				// Match the value to a label and return it
				return wordDefinitions.thousands(itemCount);
			}
		};


	/**
	* Main function
	* Process the input and translate into a word
	*/
		/*
		 * Primer
		 * This function takes an input array, reverses it (to free up the first digit) and divides it into blocks of 3.
		 * It then tries to map definitions based on the number that has and adds extra words, see below.
		 * Blocks of three because that makes it easy to translate.
		 * English translations can be thought of as descriptions for blocks of 3 numbers
		 * Aditionally, there are "in-between-labels" such as "thousand and"
		 * E.g.: "400 240" -> "fourhundred *thousand and*  twohundred fourty"
		 * This labels are passed on at the end by calculating how many numbers are after the current block.
		 */

		// 0. Filter out a 'zero'
		if (arabic === '0'){
			result.english.value = 'zero';
			if (callback){
				callback(result);
			} else return result;

			return;
		}

		// 1. Throw the entire input into an array
		let inputArray = [...arabic];

		// 2. Reverse it so that it is spaced into neat blocks and only the first section of the number has gaps in the array
		// (so that we can name it properly when we add thousands)
		inputArray = helpers.reverse(inputArray);

		// 3. Divide it into chunks and fill it to not mess up the mapping later on
		inputArray = helpers.divideInto(inputArray, 3);

		inputArray.map(function (item) {
			while (item.length != 3){
				item.push(null);
			}
		});

		// 4. Map each item to a definition
		let untranslatedArray = inputArray,
			translation = inputArray.map(helpers.matchDefinitions);

		// 5. Reverse it again
		translation = helpers.reverse(translation);

		// 6. Filter empty
		translation = translation.filter(Boolean);

		// 7. Add in-between-labels
		const translationItems = translation.length;

		translation.forEach(function (item, index) {

			let blockIndex = index,
				itemCount = 0;

			// Grab the array as it was before I translated it
			untranslatedArray.forEach(function (item, index) {
				if (index > blockIndex){
					itemCount = itemCount + untranslatedArray[index].length;
				}
			});

			// Plug in the labels
			let label = helpers.addInBetweens(itemCount);

			// If it's the last item in the array, don't add an 'and', otherwise do
			if (index + 1 == translationItems && translationItems != 1){
				//don't do it
			} else {
				label = label + 'and';
			}

			if (label != '') {
				translation[index] = item + label;
			}
		});

		// 8. Clean up
		if (translation.length == 1) {

			// If there is only 1 item in the array, you don't need the "and" part.
			translation = translation.join(' ');
			translation = translation.split('and');
			translation = translation[0];

			// Remove duplicate and trailing/leading whitespaces
			translation = mutationHelpers.typography.removeWhitespaceTrailing(translation);
			result.english.value = translation;
		} else {

			// Join the array
			translation = translation.join(' ');

			// Remove duplicate and trailing/leading whitespaces
			translation = mutationHelpers.typography.removeWhitespaceTrailing(translation);
			result.english.value = translation;
		}

		if (callback){
			callback(result);
		} else return result;
};

export default numberToWords;
